### PR TITLE
GlobalRefining: Do not refine mutable exported globals

### PR DESCRIPTION
**Pull request by kripken**
A mutable exported global might be shared with another module which
writes to it using the current type, which is unsafe and the type system does
not allow, so do not refine there.

Noticed while fuzzing wasm-split.

### CODE REVIEW

File: src/passes/GlobalRefining.cpp [hunk 1]
```diff
@@ -67,9 +67,16 @@ struct GlobalRefining : public Pass {
...
+    // another module might import our exported globals with the current type
+    // (that type is a contract between them), and in such a case the type of
+    // mutable globals must match precisely (the same rule as for mutable struct
+    // fields in subtypes - the types must match exactly, or else a write in
+    // one place could store a type considered in valid in another place).
     std::unordered_set<Name> unoptimizable;
-    if (getPassOptions().closedWorld) {
-      for (auto* global : ExportUtils::getExportedGlobals(*module)) {
+    for (auto* global : ExportUtils::getExportedGlobals(*module)) {
+      if (getPassOptions().closedWorld || global->mutable_) {
```
Comment (tlively) [inline src/passes/GlobalRefining.cpp:79]:
```suggestion
      if (!getPassOptions().closedWorld || global->mutable_) {
```
Right?

Comment (kripken) [reply src/passes/GlobalRefining.cpp:79]: I don't think so. Forget mutability for a second, then before we had
```cpp
    if (getPassOptions().closedWorld) {
      for (auto* global : ExportUtils::getExportedGlobals(*module)) {
```
and now we have
```cpp
    for (auto* global : ExportUtils::getExportedGlobals(*module)) {
      if (getPassOptions().closedWorld) {
```
which are equivalent - the check was just moved into the loop (inefficienctly).

Then there is the added check for mutability added (which is why the inefficiency seems worth it, for more concise code here).

And logically, we want to reach new line 80 (which sets unoptimizability) when closed world (so we get there for all exports), or, if open, then when mutable.

Comment (tlively) [reply src/passes/GlobalRefining.cpp:79]: Wait, but shouldn't the immutable globals be optimizable in closed world?

Comment (kripken) [reply src/passes/GlobalRefining.cpp:79]: In theory yes, but see pre-existing lines 67-69, we are strict there and avoid all exports, due to public/private issues. There is a TODO to improve that. (This PR changes nothing regarding that.)

Comment (tlively) [reply src/passes/GlobalRefining.cpp:79]: Ah, I see. Thanks!

**Review by tlively — APPROVED** [review]

**Pull request merged**